src/hashcash.cr

```diff
@@ -15,127 +14,44 @@ module Hashcash
-      @stamp_string = ""
-    )
-    end
-
-    def generate : String
-      random_string = Random::Secure.base64(12)
-
-      first_part = "#{@version}:#{@bits}:#{@date.to_s("%y%m%d%H%M%S")}:#{@resource}:#{@ext}:#{random_string}:"
-
-      counter = 0
-      stamp_string = ""
-      while stamp_string == ""
-        test_stamp = first_part + Base64.encode(counter.to_s)
-
-        # check that the first @bits bits are 0s
-        digest = Digest::SHA1.digest test_stamp
-        stamp_string = test_stamp if check digest
-
-        counter += 1
-      end
-
-      @stamp_string = stamp_string.chomp
-    end
-
-    def self.parse(stamp : String)
-      parts = stamp.split(":")
-      version, bits, date, resource, ext = parts
-      Hashcash::Stamp.new(
-        resource,
-        version.to_i,
-        bits.to_i,
-        Time.parse_utc(date, "%y%m%d%H%M%S"),
-        ext,
-        stamp
-      )
-    end
-
-    def is_for?(resource : String) : Bool
-      stamp_resource = @stamp_string.split(":")[3]
-      stamp_resource == resource
-    end
-
-    def expired?(window = 2.days.ago..2.days.from_now) : Bool
-      !window.includes?(@date)
-    end
-
-    def correct_bits?(bits = 20) : Bool
-      digest = Digest::SHA1.digest @stamp_string
-      check(digest, bits)
-    end
-
-    def valid?(
-      stamp : String,
-      resource : String,
-      time_window = 2.day.ago..2.days.from_now,
-      bits = 20
-    ) : Bool
-      parsed_stamp = Hashcash::Stamp.parse(stamp)
-      case
-      when !parsed_stamp.is_for?(resource)
-        false
-      when parsed_stamp.expired?(time_window)
-        false
-      when !parsed_stamp.correct_bits?(bits)
-        false
-      else
-        true
-      end
-    end
-
-    def self.verify!(
-      stamp : String,
-      resource : String,
-      time_window = 2.day.ago..2.days.from_now,
-      bits = 20
-    ) : Nil
-      parsed_stamp = Hashcash::Stamp.parse(stamp)
-      case
-      when !parsed_stamp.is_for?(resource)
-        raise "Stamp is not valid for the given resource(s)."
-      when parsed_stamp.expired?(time_window)
-        raise "Stamp is expired/not yet valid"
-      when !parsed_stamp.correct_bits?(bits)
-        raise "Invalid stamp, not enough 0 bits"
-      else
-        true
-      end
-    end
-
-    # TODO update method to count the number of 0s at the start rather than check it matches
-    private def check(digest : Bytes, bits = 20) : Bool
-      full_bytes = bits // 8
-      extra_bits = bits % 8
-
-      full = digest[0...full_bytes]
-      partial = digest[full_bytes]
-
-      return false unless full.all? 0
-      partial >> (8 - extra_bits) == 0
+  # TODO doc for this method
+  def self.verify!(
+    stamp_string : String,
+    resource : String,
+    time_window = 2.days.ago..2.days.from_now,
+    bits = 20
+  ) : Nil
+    stamp = Hashcash::Stamp.parse(stamp_string)
+
+    case
+    when !stamp.is_for?(resource)
+      raise "Stamp is not valid for the given resource(s)."
```

Comment:
You can set up some domain-specific exception types to encode these different failure modes. This then allows them to be handled as appropriate in user code.

For an example: https://github.com/crystal-lang/crystal/blob/5999ae29b/src/oauth2/error.cr.

Reply:
What exception types would be best to use here?

Reply:
You may want something like:

```crystal
module Hashcash
  class InvalidResource < Exception; end
  class Expired < Exception; end
  class InvalidPreimage < Exception; end
end
```